p2p/lightchain.py

```diff
@@ -245,35 +263,105 @@ def peer_block_height(peer: LESPeer) -> int:
-        return await peer.get_block_by_hash(block_hash, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_block_bodies([block_hash], request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['bodies']:
+            raise BlockNotFound("Peer {} has no block with hash {}".format(peer, block_hash))
+        return reply['bodies'][0]
 
     # TODO add a get_receipts() method to BaseChain API, and dispatch to this, as needed
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_receipts(self, block_hash: Hash32) -> List[Receipt]:
         peer = await self.get_best_peer()
         self.logger.debug("Fetching %s receipts from %s", encode_hex(block_hash), peer)
-        return await peer.get_receipts(block_hash, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_receipts(block_hash, request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['receipts']:
+            raise BlockNotFound("No block with hash {} found".format(block_hash))
+        return reply['receipts'][0]
 
     # TODO implement AccountDB exceptions that provide the info needed to
     # request accounts and code (and storage?)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_account(self, block_hash: Hash32, address: Address) -> Account:
         peer = await self.get_best_peer()
-        return await peer.get_account(block_hash, address, self.cancel_token)
+        key = keccak(address)
+        proof = await self._get_proof(peer, block_hash, account_key=b'', key=key)
+        header = await self._get_block_header_by_hash(peer, block_hash)
+        rlp_account = HexaryTrie.get_from_proof(header.state_root, key, proof)
+        return rlp.decode(rlp_account, sedes=Account)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_contract_code(self, block_hash: Hash32, key: bytes) -> bytes:
         peer = await self.get_best_peer()
-        return await peer.get_contract_code(block_hash, key, self.cancel_token)
+        request_id = gen_request_id()
+        peer.sub_proto.send_get_contract_code(block_hash, key, request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['codes']:
+            return b''
+        return reply['codes'][0]
+
+    async def _get_block_header_by_hash(self, peer: LESPeer, block_hash: Hash32) -> BlockHeader:
+        self.logger.debug("Fetching header %s from %s", encode_hex(block_hash), peer)
+        request_id = gen_request_id()
+        max_headers = 1
+        peer.sub_proto.send_get_block_headers(block_hash, max_headers, request_id)
+        reply = await self._wait_for_reply(request_id)
+        if not reply['headers']:
+            raise HeaderNotFound("Peer {} has no block with hash {}".format(peer, block_hash))
+        return reply['headers'][0]
```

Comment:
Same here, I expected to see a check that the returned `header.hash` actually matches the requested `block_hash`

Reply:
I've added a check raising an exception if the received header's hash does not match `block_hash`